### PR TITLE
add crd for pulsar package & function & connector

### PR DESCRIPTION
**Pull request by freeznet**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by nlu90 — APPROVED** [review]

**Comment by ericsyh** [comment]
@labuladong Pls take a look this PR.

**Comment by nlu90** [comment]
@labuladong could you help review this PR ASAP?

File: pkg/connection/reconcile_function.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		return ""
+	}
+	if lowerURL := strings.ToLower(url); strings.HasPrefix(lowerURL, "http://") ||
+		strings.HasPrefix(lowerURL, "https://") ||
+		strings.HasPrefix(lowerURL, "builtin://") ||
+		strings.HasPrefix(lowerURL, "file://") {
+		return url
+	}
+	if _, err := utils.GetPackageName(url); err != nil {
+		return ""
```
Comment (labuladong) [inline pkg/connection/reconcile_function.go:170]: suggest to return the error too

**Review by nlu90 — APPROVED** [review]

File: pkg/connection/reconciler.go [hunk 2]
```diff
@@ -67,6 +74,10 @@ func MakeReconciler(log logr.Logger, k8sClient client.Client, creator admin.Puls
 		makeNamespacesReconciler(r),
 		makeTopicsReconciler(r),
 		makePermissionsReconciler(r),
+		makeFunctionsReconciler(r),
+		makeSinksReconciler(r),
+		makeSourcesReconciler(r),
+		makePackagesReconciler(r),
```
Comment (labuladong) [inline pkg/connection/reconciler.go:80]: https://github.com/streamnative/pulsar-resources-operator/blob/087c81ebe0c52c38d5e6006f2d761ba87b477b74/pkg/connection/reconciler.go#L176-L181

These reconcilers will execute one by one, one fail will block the flowing reconcilers. So the order is important, I think the function reconciler relies on package reconciler? What will happen if you create function resource first, then create package resource?

**Comment by labuladong** [comment]
I'm reviewing, I think there are some problems in this pr

File: pkg/connection/reconciler.go (hunk 2 above)
Comment (labuladong) [reply pkg/connection/reconciler.go:80]: fix in https://github.com/streamnative/pulsar-resources-operator/pull/216

**Pull request merged**